client/client-lib/src/query.rs

```diff
@@ -0,0 +1,194 @@
+                QueryStep::Finished(Ok(epoch))
+            }
+            response => self.current.process(response, max_evil),
+        }
+    }
+}
+
+/// Returns the deduplicated union of f+1 responses
+pub struct UnionResponses<R> {
+    responses: HashSet<PeerId>,
+    results: HashSet<R>,
+    current: CurrentConsensus<Vec<R>>,
+}
+
+impl<R> Default for UnionResponses<R> {
+    fn default() -> Self {
+        Self {
+            responses: HashSet::new(),
+            results: HashSet::new(),
+            current: CurrentConsensus::default(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<Vec<R>> for UnionResponses<R> {
+    fn process(&mut self, response: FedResponse<Vec<R>>, max_evil: usize) -> QueryStep<Vec<R>> {
+        return if let (peer, Ok(result)) = response {
+            self.results.extend(result.into_iter());
+            self.responses.insert(peer);
+
+            if self.responses.len() > max_evil {
+                QueryStep::Finished(Ok(self.results.drain().collect()))
+            } else {
+                QueryStep::Continue
+            }
+        } else {
+            // handle error case using the CurrentConsensus method
+            self.current.process(response, max_evil)
+        };
+    }
+}
+
+/// Returns when f+1 responses are equal, retrying on 404 errors
+pub struct Retry404<R> {
+    current: CurrentConsensus<R>,
+}
+
+impl<R> Default for Retry404<R> {
+    fn default() -> Self {
+        Self {
+            current: CurrentConsensus::default(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for Retry404<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        match response {
+            (peer, Err(JsonRpcError::Call(RpcCallError::Custom(e)))) if e.code() == 404 => {
+                QueryStep::Request(HashSet::from([peer]))
+            }
+            response => self.current.process(response, max_evil),
+        }
+    }
+}
+
+/// Returns when f+1 responses are equal, retrying after every 2f+1 responses
+pub struct EventuallyConsistent<R> {
+    responses: HashSet<PeerId>,
+    current: CurrentConsensus<R>,
+}
+
+impl<R> Default for EventuallyConsistent<R> {
+    fn default() -> Self {
+        Self {
+            responses: HashSet::new(),
+            current: CurrentConsensus::default(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for EventuallyConsistent<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        let honest_threshold = max_evil * 2 + 1;
+        self.responses.insert(response.0);
+
+        match self.current.process(response, max_evil) {
+            QueryStep::Continue if self.responses.len() == honest_threshold => {
+                let result = QueryStep::Request(self.responses.clone());
+                self.responses.clear();
+                result
+            }
```

Comment:
I don't think there's an advantage to sending a new request to a peer that hasn't responded yet.

The epoch-aware API introduces more complexity since we'll have to store that information in the database, luckily it's unlikely in practice, at least for now I'll add the FIXME.